src/TestExplorer/LSPTestDiscovery.ts

```diff
@@ -29,104 +34,118 @@ import { LanguageClientManager } from "../sourcekit-lsp/LanguageClientManager";
+                    { textDocument: { uri: document.toString() } },
+                    token
+                );
+                return this.transform(client, swiftPackage, testsInDocument);
+            } else {
+                throw new Error(`${textDocumentTestsRequest.method} requests not supported`);
+            }
+        });
     }
 
     /**
      * Return list of workspace tests
      * @param workspaceRoot Root of current workspace folder
      */
-    async getWorkspaceTests(workspaceRoot: vscode.Uri): Promise<TestDiscovery.TestClass[]> {
+    async getWorkspaceTests(
+        swiftPackage: SwiftPackage,
+        workspaceRoot: vscode.Uri
+    ): Promise<TestDiscovery.TestClass[]> {
         return await this.languageClient.useLanguageClient(async (client, token) => {
-            const tests = await client.sendRequest(workspaceTestsRequest, {}, token);
-            const testsInWorkspace = tests.filter(item =>
-                isPathInsidePath(
-                    client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                    workspaceRoot.fsPath
-                )
-            );
-            const classes = testsInWorkspace
-                .filter(item => {
-                    return (
-                        item.kind === langclient.SymbolKind.Class &&
-                        isPathInsidePath(
-                            client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                            workspaceRoot.fsPath
-                        )
-                    );
-                })
-                .map(item => {
-                    const functions = testsInWorkspace
-                        .filter(func => func.containerName === item.name)
-                        .map(func => {
-                            const openBrackets = func.name.indexOf("(");
-                            let funcName = func.name;
-                            if (openBrackets) {
-                                funcName = func.name.slice(0, openBrackets);
-                            }
-                            return {
-                                name: funcName,
-                                location: client.protocol2CodeConverter.asLocation(func.location),
-                            };
-                        });
-                    return {
-                        name: item.name,
-                        location: client.protocol2CodeConverter.asLocation(item.location),
-                        functions: functions,
-                    };
-                });
-            return classes;
+            // Only use the lsp for this request if it supports the
+            // workspace/tests method, and is at least version 2.
+            if (this.checkExperimentalCapability(client, workspaceTestsRequest.method, 2)) {
+                const tests = await client.sendRequest(workspaceTestsRequest, {}, token);
+                const testsInWorkspace = tests.filter(item =>
```

Comment:
Have you found a situation where the tests returned weren't in the workspace? Also `updateTestsFromClasses` does this for individual workspace folders later on

Reply:
I think you're right. I kept this filter from the old code but I dont think the LSP can return tests outside the workspace it was initialized with. I'll remove it.